File: backend/app/routes/organizations.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from uuid import UUID
from app.database import get_db
from app.schemas import Organization, OrganizationCreate, OrganizationUpdate, CognitoUser
from app.services.organization_service import OrganizationService
from app.middleware.auth import get_current_user, require_role

router = APIRouter(prefix="/api/organizations", tags=["organizations"])


@router.post("/", response_model=Organization, status_code=status.HTTP_201_CREATED)
async def create_organization(
    org_data: OrganizationCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CognitoUser = Depends(require_role("system_admin"))
):
    """Create a new organization (system_admin only)"""
    return await OrganizationService.create(db, org_data)


@router.get("/", response_model=list[Organization])
async def list_organizations(
    db: AsyncSession = Depends(get_db),
    current_user: CognitoUser = Depends(require_role("system_admin"))
):
    """List all organizations (system_admin only)"""
    return await OrganizationService.list_all(db)


@router.get("/{org_id}", response_model=Organization)
async def get_organization(
    org_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: CognitoUser = Depends(get_current_user)
):
    """Get organization by ID"""
    # Users can only view their own organization unless they're system_admin
    if str(org_id) != current_user.org_id and current_user.role != "system_admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied"
        )
    
    org = await OrganizationService.get_by_id(db, org_id)
    if not org:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Organization not found"
        )
    
    return org


@router.patch("/{org_id}", response_model=Organization)
async def update_organization(
    org_id: UUID,
    org_data: OrganizationUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CognitoUser = Depends(require_role("system_admin"))
):
    """Update organization (system_admin only)"""
    org = await OrganizationService.update(db, org_id, org_data)
    if not org:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Organization not found"
        )
    
    return org


@router.delete("/{org_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_organization(
    org_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: CognitoUser = Depends(require_role("system_admin"))
):
    """Delete organization (system_admin only)"""
    deleted = await OrganizationService.delete(db, org_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Organization not found"
        )

File: backend/src/routes/organizations.ts
import { Router } from 'express';
import { OrganizationService } from '../services/organization.service';
import { authenticateToken, requireRole } from '../middleware/auth';
import { AuthRequest } from '../types';

const router = Router();
const orgService = new OrganizationService();

// All routes require authentication
router.use(authenticateToken);

// Create organization (system_admin only)
router.post(
  '/',
  requireRole('system_admin'),
  async (req: AuthRequest, res) => {
    try {
      const { name, type } = req.body;

      if (!name || !type) {
        return res.status(400).json({ error: 'Name and type are required' });
      }

      if (!['police', 'medical'].includes(type)) {
        return res.status(400).json({ error: 'Type must be police or medical' });
      }

      const org = await orgService.createOrganization(name, type);
      res.status(201).json(org);
    } catch (error) {
      console.error('Create organization error:', error);
      res.status(500).json({ error: 'Failed to create organization' });
    }
  }
);

// List all organizations (system_admin only)
router.get(
  '/',
  requireRole('system_admin'),
  async (req: AuthRequest, res) => {
    try {
      const orgs = await orgService.listOrganizations();
      res.json(orgs);
    } catch (error) {
      console.error('List organizations error:', error);
      res.status(500).json({ error: 'Failed to list organizations' });
    }
  }
);

// Get organization by ID
router.get('/:id', async (req: AuthRequest, res) => {
  try {
    const { id } = req.params;
    const userOrgId = req.user?.['custom:org_id'];

    // Users can only view their own organization unless they're system_admin
    if (userOrgId !== id && req.user?.['custom:role'] !== 'system_admin') {
      return res.status(403).json({ error: 'Access denied' });
    }

    const org = await orgService.getOrganization(id);
    
    if (!org) {
      return res.status(404).json({ error: 'Organization not found' });
    }

    res.json(org);
  } catch (error) {
    console.error('Get organization error:', error);
    res.status(500).json({ error: 'Failed to get organization' });
  }
});

// Update organization (system_admin only)
router.patch(
  '/:id',
  requireRole('system_admin'),
  async (req: AuthRequest, res) => {
    try {
      const { id } = req.params;
      const updates = req.body;

      const org = await orgService.updateOrganization(id, updates);
      
      if (!org) {
        return res.status(404).json({ error: 'Organization not found' });
      }

      res.json(org);
    } catch (error) {
      console.error('Update organization error:', error);
      res.status(500).json({ error: 'Failed to update organization' });
    }
  }
);

// Delete organization (system_admin only)
router.delete(
  '/:id',
  requireRole('system_admin'),
  async (req: AuthRequest, res) => {
    try {
      const { id } = req.params;
      const deleted = await orgService.deleteOrganization(id);
      
      if (!deleted) {
        return res.status(404).json({ error: 'Organization not found' });
      }

      res.status(204).send();
    } catch (error) {
      console.error('Delete organization error:', error);
      res.status(500).json({ error: 'Failed to delete organization' });
    }
  }
);

export default router;
